packages/auth/lib/CSRFProtection.ts

```diff
@@ -0,0 +1,73 @@
+import { sha256 } from "./crypto";
+import { URL } from "url";
+import { ILogger } from "@fullstack-one/logger";
+
+export class CSRFProtection {
+  private logger: ILogger;
+  private authConfig;
+
+  constructor(logger: ILogger, authConfig) {
+    this.logger = logger;
+    this.authConfig = authConfig;
+  }
+
+  public async createSecurityContext(ctx, next) {
+    ctx.securityContext = {
+      isBrowser: true,
+      isApiClient: false,
+      clientIdentifier: null,
+      requestMethod: ctx.request.method.toUpperCase(),
+      sameOriginApproved: {
+        byReferrer: false,
+        byOrigin: false
+      }
+    };
+
+    // Generate clientIdentifier for refresh-token
+    if (ctx.request.ip != null && ctx.request.headers["user-agent"] != null) {
+      ctx.securityContext.clientIdentifier = sha256(`${ctx.request.ip}_#_${ctx.request.headers["user-agent"]}`);
+    }
+
+    // Check if https is used on production
+    if (process.env.NODE_ENV === "production") {
+      if (this.authConfig.enforceHttpsOnProduction !== false && ctx.request.protocol !== "https") {
+        this.logger.warn("Request rejected: Unsecure requests are not allowed here. Please use HTTPS.", ctx.securityContext);
+        // return ctx.throw(400, "Unsecure requests are not allowed here. Please use HTTPS.");
+      }
+    }
+
+    const origin = ctx.request.get("origin");
+    const referrer = ctx.request.get("referrer");
+    let referrerOrigin = null;
+
+    // Validate same origin policy by origin header
```

Comment:
We could remove the comment if we had a separate function

Reply:
Let's discuss in the comment above: https://github.com/fullstack-build/fullstack-one/pull/48/files/7b9f3b5d08da7648fe71e119d860a8c460424b0b#r275121751